### PR TITLE
Bump protobuf to 3.10.0 and apply fixes for --incompatible_load_{java|proto|python}_rules_from_bzl

### PR DESCRIPTION
**Pull request by davido**
The JS tests are failing:

```
  $ bazel test closure/...
ERROR: /home/davido/projects/davido_rules_closure/closure/templates/test/BUILD:105:1: Compiling 94 JavaScript files to closure/templates/test/greeter_proto_test_bin.js failed (Exit 1)
external/com_google_protobuf_js/binary/writer.js:239: ERROR - Bad type annotation. Unknown type goog.crypt.base64.Alphabet
 * @param {!goog.crypt.base64.Alphabet=} alphabet Which flavor of base64 to use.
            ^
  ProTip: "JSC_UNRECOGNIZED_TYPE_ERROR" or "checkTypes" or "unrecognizedTypeError" can be added to the `suppress` attribute of:
  //closure/protobuf:jspb
  Alternatively /** @suppress {checkTypes} */ can be added to the source file.

1 error(s), 0 warning(s), 97.2% typed
	
INFO: Elapsed time: 91.160s, Critical Path: 44.83s
INFO: 1074 processes: 164 linux-sandbox, 910 worker.
FAILED: Build did NOT complete successfully
```

That is defined here: [1] and that was added here: [2]. It seems that closure compiler has to be bumped as well.

[1] https://github.com/google/closure-library/blame/master/closure/goog/crypt/base64.js#L71
[2] https://github.com/google/closure-library/releases/tag/v20190729

### CODE REVIEW

**Comment by davido** [comment]
//CC @Yannic @gkdn

File: closure/repositories.bzl [hunk 1]
```diff
@@ -69,6 +69,8 @@ def closure_repositories(
         omit_org_ow2_asm_tree = False,
         omit_org_ow2_asm_util = False,
         omit_phantomjs = False,
+        omit_rules_proto = False,
+        omit_rules_python = False,
```
Comment (Yannic) [inline closure/repositories.bzl]: Please also include `rules_{cc,java}`. They are required by the native rules, so they are implicit dependencies, but I'm not sure how long they'll stay that way.

File: java/io/bazel/rules/closure/common/BUILD~ [hunk 2]
```diff
@@ -0,0 +1,13 @@
+java_library(
```
Comment (Yannic) [inline java/io/bazel/rules/closure/common/BUILD~]: Why is this added?

File: java/io/bazel/rules/closure/common/CheckReturnValue.java~ [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2016 The Closure Rules Authors. All Rights Reserved.
```
Comment (Yannic) [inline java/io/bazel/rules/closure/common/CheckReturnValue.java~]: Why is this added (dito `*.java~`)?

File: closure/library/BUILD [hunk 4]
```diff
@@ -1,11 +1,12 @@
 # DO NOT EDIT -- bazel run //closure/library:regenerate -- "$PWD"
 
+load("@rules_python//python:defs.bzl", "py_binary")
```
[the remark this answers is not in the document]
Comment (davido) [reply closure/library/BUILD]: Done

File: closure/repositories.bzl [hunk 5]
```diff
@@ -999,6 +1005,22 @@ def phantomjs():
...
+    http_archive(
+        name = "rules_python",
+        sha256 = "e5470e92a18aa51830db99a4d9c492cc613761d5bdb7131c04bd92b9834380f6",
+        strip_prefix = "rules_python-4b84ad270387a7c439ebdccfd530e2339601ef27",
+        urls = ["https://github.com/bazelbuild/rules_python/archive/4b84ad270387a7c439ebdccfd530e2339601ef27.tar.gz"],
+    )
+
 def zlib():
     http_archive(
         name = "zlib",
```
[the remark this answers is not in the document]
Comment (davido) [reply closure/repositories.bzl]: Done.

File: java/io/bazel/rules/closure/common/BUILD~ (hunk 2 above)
Comment (davido) [reply java/io/bazel/rules/closure/common/BUILD~]: Done.

File: java/io/bazel/rules/closure/common/CheckReturnValue.java~ (hunk 3 above)
Comment (davido) [reply java/io/bazel/rules/closure/common/CheckReturnValue.java~]: Done.

File: closure/repositories.bzl (hunk 1 above)
Comment (davido) [reply closure/repositories.bzl]: I would rather not add them in this PR, but prefer to add the later. In Gerrit we have also consumed Java rules from Starlark, and not fetched the rules in WORKSPACE file.

Comment (Yannic) [reply closure/repositories.bzl]: OK, I think they'll be available at least during `1.x`, so that shouldn't be an issue for the next 3+ month.

Comment (Yannic) [reply closure/repositories.bzl]: Actually, I think `rules_{cc,java}` is only available in Bazel 0.29+. Please add them in this PR.

With Bazel 0.28.1, I'm seeing the following error:

```
ERROR: error loading package 'closure/templates': Unable to find package for @rules_java//java:defs.bzl: The repository '@rules_java' could not be resolved.
INFO: Elapsed time: 0.211s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (3 packages loaded)
```

Comment (davido) [reply closure/repositories.bzl]: I've added `rules_{cc,java}`. Done.

**Comment by davido** [comment]
@Yannic @gkdn @dslomov @laurentlb 

I rebased this PR on recent master and squashed all commits. PTAL.

**Comment by davido** [comment]
OK, there is one complication. I conducted custom `rules_closure` release from this commit here: [1], and tried to consume this new rules_closure release in Gerrit project: [2]. The build is broken, and it seems that the same problem shows up that I already reported here (but that I was able to workaround in Gerrit build tool chain, by moving java_import invocation from Starlark file to BUILD file): [3]:

```
  $ bazel build :release
INFO: Writing tracer profile to '/home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/command.profile.gz'
INFO: Invocation ID: 6cb4e17d-97db-4bff-b824-5342fe18a6b2
ERROR: Failed to load Starlark extension '@rules_java//java:defs.bzl'.
Cycle in the workspace file detected. This indicates that a repository is used prior to being defined.
The following chain of repository dependencies lead to the missing definition.
 - @rules_java
This could either mean you have to add the '@rules_java' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or move an existing definition earlier in your WORKSPACE file.
ERROR: cycles detected during target parsing
INFO: Elapsed time: 0.489s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```

Bazel version is Bazel@HEAD, built from this commit (6470bb74047a76d3e0271273ac0864b697ab2aa9).

[1] https://github.com/davido/rules_closure/releases/tag/v0.25
[2] https://gerrit-review.googlesource.com/c/gerrit/+/237186
[3] https://github.com/bazelbuild/bazel/issues/9291

**Comment by Yannic** [comment]
@davido Ok, I was able to reproduce that locally with Gerrit. I think it's caused by the fact that we tell users to use `load("@io_bazel_rules_closure//closure:defs.bzl", "closure_repositories")` which also exports our rules, hence has load statements for repos we haven't declared yet (and is against Bazel's [rule guide](https://docs.bazel.build/versions/master/skylark/deploying.html)).

The reason this doesn't fail on our CI is that we use `load("@io_bazel_rules_closure//closure:repositories.bzl", "closure_repositories")`. Switching Gerrit to that one also seems to fix Gerrit.

Fixing this is going to be a bit of a pain because it's an incompatible change.
I propose the following migration plan:
  - Create `rules_proto_{dependencies,toolchains}` in `closure/repositories.bzl` to conform with the rule guide (and update readme).
  - Make `closure_repositories` in `closure/defs.bzl` a macro that prints a warning to tell users to switch to `rules_proto_{dependencies,toolchains}`.
  - Cut a release.
  - Land this PR, Gerrit can consume rules_closure at head.
  - `2019-11-01`: Remove `closure_repositories` from `closure/defs.bzl` and cut a new release.

**Comment by davido** [comment]
@Yannic, I added this [PR](https://github.com/bazelbuild/rules_closure/pull/425), where I refactored the consumption of rules_closure dependencies.

**Comment by davido** [comment]
Can this now be now merged, or do we waiting for something?

**Comment by Yannic** [comment]
@davido Can you rebase to resolve the merge conflict and update Travis to `0.29.1`? I think that's all that's blocking us from merging.

@gdkn @dslomov Please cut a new release before merging this. There is a chance that this breaks users if they don't explicitly declare `rules_{cc,java}` (but merging this is the way forward).

**Comment by davido** [comment]
> @davido Can you rebase to resolve the merge conflict and update Travis to 0.29.1?

Done.

**Comment by davido** [comment]
@Yannic The build on Mac OS slave is failing:

```
ERROR: /Users/travis/build/bazelbuild/rules_closure/closure/compiler/test/closure_js_deps/BUILD:66:1: Calculating 39 JavaScript deps to closure/compiler/test/closure_js_deps/goblin_deps.js failed (Exit 1) depswriter failed: error executing command 
  (cd /Users/travis/.cache/bazel/sandbox/darwin-sandbox/289/execroot/io_bazel_rules_closure && \
  exec env - \
  bazel-out/host/bin/external/com_google_javascript_closure_library/depswriter '--output_file=bazel-out/darwin-fastbuild/bin/closure/compiler/test/closure_js_deps/goblin_deps.js' '--root_with_prefix=closure/compiler/test/closure_js_deps ../../../io_bazel_rules_closure/closure/compiler/test/closure_js_deps' '--root_with_prefix=external/com_google_javascript_closure_library/closure/goog .')
Execution platform: @bazel_tools//platforms:host_platform
Use --sandbox_debug to see verbose messages from the sandbox
[365 / 370] .../compiler/test/closure_js_deps:hyperion2_lib; 0s darwin-sandbox
pting to use the default Python toolchain (@rules_python//python:autodetecting_toolchain).
According to '/usr/local/bin/python -V', version is 'Python 2.7.12', but we need version 3. PATH is:
/usr/gnu/bin:/usr/local/bin:/bin:/usr/bin:.
Please ensure an interpreter with version 3 is available on this platform as 'python3' or 'python', or else register an appropriate Python toolchain as per the documentation for py_runtime_pair (https://github.com/bazelbuild/rules_python/blob/master/docs/python.md#py_runtime_pair).
Note that prior to Bazel 0.27, there was no check to ensure that the interpreter's version matched the version declared by the target (#4815). If your build worked prior to Bazel 0.27, and you're sure your targets do not require Python 3, you can opt out of this version check by using the non-strict autodetecting toolchain instead of the standard autodetecting toolchain. This can be done by passing the flag `--extra_toolchains=@rules_python//python:autodetecting_toolchain_nonstrict` on the command line or adding it to your bazelrc.
----------------
Note: The failure of target @com_google_javascript_closure_library//:depswriter (with exit code 1) may have been caused by the fact that it is running under Python 3 instead of Python 2. Examine the error to determine if that appears to be the problem. Since this target is built in the host configuration, the only way to change its version is to set --host_force_python=PY2, which affects the entire build.
If this error started occurring in Bazel 0.27 and later, it may be because the Python toolchain now enforces that targets analyzed as PY2 and PY3 run under a Python 2 and Python 3 interpreter, respectively. See https://github.com/bazelbuild/bazel/issues/7899 for more information.
```

I remember we has a similar issue in gerrit and solution was to add the following options in `.bazelrc`: 

```
build --action_env=PATH
build --incompatible_strict_action_env
```

**Comment by davido** [comment]
@brandjon @Yannic 

I added `.bazelrc` wiht the following content:

```
build --action_env=PATH
build --incompatible_strict_action_env
```

Let's see, what happens.

**Comment by Yannic** [comment]
Yeah, Bazel 0.25 changed the default version of Python, which I think is why we're seeing this.

I'd rather not set any `--incompatible_*` flag. Feel free to try setting `lang: python3`.
https://github.com/bazelbuild/rules_closure/pull/422/files#diff-354f30a63fb0907d4ad57269548329e3R6

**Comment by davido** [comment]
> Yeah, Bazel 0.25 changed the default version of Python, which I think is why we're seeing this.

Ah, OK. I removed `.bazelrc` again and added Python 3 to Travis instead, as suggested.

**Comment by brandjon** [comment]
@davido's error log indicates that your build is trying to run a py_binary (depswriter) in the host config. By default, the host config uses PY3 (regardless of python_version). This fails at execution time because your machine apparently does not have a python3 interpreter.

So presumably, you don't want to be using Python 3 in your build. You can change the default version used by the host config to PY2 by passing --host_force_python=PY2 on the command line.

If for some reason you can't do that -- for instance, if you have targets that depend on Bazel thinking that they're PY3 (in `select()`s) when they're really not -- you can opt into the old behavior of not enforcing the declared Python version at execution time by using the nonstrict toolchain, as mentioned in the error message. I'd avoid that.

**Comment by davido** [comment]
@brandjon Thanks for your comment. Even after adding "language: python3" the Python 3 is still not found on Mac OS; just check the very last log.

I remember one issue in Gerrit where `build --action_env=PATH` had to be added to the `.bazelrc` file to detect Python 3 on Mac OS on Bazel-CI. I just added `.bazelrc` with that content.

**Comment by Yannic** [comment]
Sorry, my bad. The correct way to declare Python 3 is `language: python`. It seems like 3.6 is the default version for Travis.
https://docs.travis-ci.com/user/languages/python/

If this also doesn't work, `--host_force_python=PY2` is an alternative. There are no `select()`s that assume PY3.

**Comment by davido** [comment]
> The correct way to declare Python 3 is language: python

OK, changed.

**Comment by Yannic** [comment]
Sorry for the repeated changes, fixing Travis really isn't fun. Please also remove `dist: trusty` (Ubuntu 14.04 is EOL, default is 16.04) and `osx_image: xcode8` (macOS 10.11 doesn't have py3.6, default should have it).

**Comment by brandjon** [comment]
Er, my previous answer was predicated on the assumption that this target is written for PY2 and will only run under PY2. If you actually want it to run under PY3, then presumably the CI config just needs to ensure Python 3 is available. Note that `/usr/local/bin` is already on PATH according to the error log, so if `python3` is installed it should find it.

**Comment by davido** [comment]
@brandjon @Yannic Thanks! I addressed all comments. Hopefully, the next build should succeed.

**Comment by Yannic** [comment]
Ok, we're at the point where I'm out of ideas how to make Travis happy. If you want to remove `osx`, I won't complain. Buildkite is covering it.

Thank you for your patience!

**Comment by Yannic** [comment]
@gkdn @dslomov What's the status on this? Can you cut a release and then merge this?

File: closure/repositories.bzl [hunk 6]
```diff
@@ -726,23 +738,23 @@ def com_google_jsinterop_annotations():
...
     http_archive(
         name = "com_google_protobuf",
-        strip_prefix = "protobuf-3.9.0",
-        sha256 = "2ee9dcec820352671eb83e081295ba43f7a4157181dad549024d7070d079cf65",
+        strip_prefix = "protobuf-3.10.0-rc1",
+        sha256 = "690ea9b2ebaa788fbf2cf3353bba2e75bd99e16b09f5e69fdb664f679e5bc584",
         urls = [
-            "https://mirror.bazel.build/github.com/google/protobuf/archive/v3.9.0.tar.gz",
-            "https://github.com/protocolbuffers/protobuf/archive/v3.9.0.tar.gz",
+            #            "https://mirror.bazel.build/github.com/google/protobuf/archive/v3.10.0-rc1.tar.gz",
```
Comment (davido) [inline closure/repositories.bzl]: The commented out mirror URL can be improved here and in other place.

@meteorcloudy @laurentlb Can you please upload `https://github.com/protocolbuffers/protobuf/archive/v3.10.0-rc1.tar.gz` to https://mirror.bazel.build/github.com/google/protobuf/archive/v3.10.0-rc1.tar.gz

Comment (rickypai) [reply closure/repositories.bzl]: it seems like `3.10.0` was just released yesterday. maybe the `-rc1` suffix can be dropped?

Comment (rickypai) [reply closure/repositories.bzl]: thank you for working on this btw!

Comment (davido) [reply closure/repositories.bzl]: Thanks for letting me know, I will amend this series and bump right to 3.10.0.

@meteorcloudy @laurentlb Could you please upload https://github.com/protocolbuffers/protobuf/archive/v3.10.0.tar.gz to bazel mirror? Thank you in advance!

Comment (meteorcloudy) [reply closure/repositories.bzl]: Done, https://mirror.bazel.build/github.com/google/protobuf/archive/v3.10.0.tar.gz and https://mirror.bazel.build/github.com/google/protobuf/archive/v3.10.0.zip are both available.

Comment (davido) [reply closure/repositories.bzl]: Thanks to everyone for help!

@gkdn @laurentlb @dslomov PTAL, it's ready now.

File: closure/library/BUILD [hunk 7]
```diff
@@ -1,11 +1,12 @@
...
+
 package(default_visibility = ["//visibility:public"])
 
 licenses(["notice"])
 
-load("//closure:defs.bzl", "closure_css_library")
-load("//closure:defs.bzl", "closure_js_library")
+load("//closure:defs.bzl", "closure_css_library", "closure_js_library")
```
Comment (gkdn) [inline closure/library/BUILD]: pls move next to other load

File: closure/protobuf/closure_proto_library.bzl [hunk 8]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@rules_proto//proto:defs.bzl", "ProtoInfo")
```
Comment (gkdn) [inline closure/protobuf/closure_proto_library.bzl]: pls keep loads together. and for others as well

File: .travis.yml [hunk 9]
```diff
@@ -1,17 +1,10 @@
...
-dist: trusty
 sudo: required
-# xcode8 has jdk8
-osx_image: xcode8
-# Not technically required but suppresses 'Ruby' in Job status message.
-language: java
 
 os:
   - linux
-  - osx
```
Comment (gkdn) [inline .travis.yml]: no more osx? I don't have the context.

File: closure/library/BUILD (hunk 7 above)
Comment (davido) [reply closure/library/BUILD]: Done.

File: closure/protobuf/closure_proto_library.bzl (hunk 8 above)
Comment (davido) [reply closure/protobuf/closure_proto_library.bzl]: Done.

File: .travis.yml (hunk 9 above)
Comment (davido) [reply .travis.yml]: Are these statements in the commit message not enough?

```
This change also does the following:
[...]
* Bump Bazel version to 0.29.1 in .travis.yml file
[...]
Travis changes:
    
    * Remove dist: trusty (Ubuntu 14.04 is EOL, default is 16.04)
    * Remove osx OS as it currently doesn't support Python: [1].

[1] https://github.com/travis-ci/travis-ci/issues/2312
```

Comment (Yannic) [reply .travis.yml]: Yeah, it's verified by Buildkite (and Travis is super slow), so there's no harm in removing it.

See also https://github.com/bazelbuild/rules_closure/pull/422#issuecomment-531787668

**Comment by davido** [comment]
@laurentlb Can you please approve this PR?

**Comment by gkdn** [comment]
I merged the patch. I also cut a new release before merging.

**Comment by davido** [comment]
Thanks!

**Comment by Yannic** [comment]
Thanks! I'll send a follow-up of #423 to update the readme / getting started section tomorrow.
